### ISSUE TITLE
18.02.03 - 《Python 机器学习及实践》笔记

### ISSUE THREAD

**Issue by bighuang624** [bighuang624/bighuang624.github.io#62]
一本入门级别的好书，主要是代码多，算法没有怎么深入。因此只记一些感觉有必要的东西

**Comment by bighuang624** [comment]
## 分类学习

### 支持向量机

**支持向量机分类器（Support Vector Classifier）**，便是根据训练样本的分布，搜索所有可能的线性分类器中最佳的那个。因为决定直线位置的并非所有的训练数据，而是其中的两个空间间隔最小的两个不同类别的数据点，这种可以用来真正帮助决策最优线性分类模型的数据点叫做“支持向量”。

### 决策树

使用多种不同特征组合搭建多层决策树的情况下，模型在学习的时候就需要**考虑特征节点的选取顺序**。常用的度量方式包括**信息熵（Information Gain）**和**基尼不纯性（Gini Impurity）**。默认配置的决策树模型使用的是基尼不纯性作为特征排序的度量指标。

**优点**：

1. 树模型可以解决**非线性特征**的问题；
2. 树模型**不要求对特征标准化和统一量化**，即数值型和类别型特征都可以直接被应用在树模型的构建和预测过程中；
3. 因为上述原因，树模型也可以直观地输出决策过程，使得**预测结果具有可解释性**。

**缺点**：

1. 正是因为树模型可以解决复杂的非线性拟合问题，所以更加容易因为模型搭建过于复杂而**降低泛化能力**；
2. 树模型从上至下的预测流程会因为数据细微的更改而发生较大的结构变化，因此**预测稳定性较差**；
3. 依托训练数据构建最佳的树模型是 NP 问题（在有限时间内无法找到最优解的问题），因此我们所使用类似贪婪算法的解法**只能找到一些次优解**，这也是为什么我们经常借助集成模型，在多个次优解中寻觅更高的模型性能。

### 集成模型（分类）

**集成（Ensemble）分类模型**是综合考量多个分类器的预测效果，从而作出决策。“综合考量”的方式大体分为以下两种：

1. 利用相同的训练数据同时搭建多个独立的分类模型，然后通过投票的方式，以少数服从多数的原则作出最终的分类决策。代表模型是**随机森林分类器（Random Forest Classifier）**，即在相同训练数据上同时搭建多棵决策树。
2. 按照一定次序搭建多个分类模型。这些模型之间彼此存在依赖关系。一般而言，每一个后续模型的加入都需要对现有集成模型的综合性能有所贡献，进而不断提升更新过后的集成模型的性能，并最终期望借助整合多个分类能力较弱的分类器，搭建出具有更强分类能力的模型。代表模型是梯度提升决策树（Gradient Tree Boosting）。

**基线系统（Baseline System）**通常指使用经典模型搭建的机器学习系统。研发人员每提出一个新模型，都需要和基线系统在多个具有代表性的数据集上进行性能比较的测试。一般而言，随机森林分类模型经常作为科研论文、数据竞赛、工业界的基线系统。

## 回归预测

### 评价指标

p68

* 平均绝对误差（Mean Absolute Error，MAE）
* 均方误差（Mean Squared Error，MSE）
* R-squared 

### 核函数

支持向量机模型可以通过配置不同的核函数来改变模型性能。**核函数**是一项非常有用的特征映射技巧，简单来说，便是通过某种函数计算，**将原有的特征映射到更高维度的空间**，从而尽可能达到新的**高维度特征线性可分**的程度。

### 极端随机森林

随机森林模型有另一个变种：极端随机森林（Extremely Randomized Trees）。与普通的随机森林模型不同的是，极端随机森林在每当构建一棵树的分裂结点时，不会任意选取特征；而是先随机搜集一部分特征，然后利用信息熵和基尼不纯性等指标挑选最佳的节点特征。

### 选择适当的模型

[Scikit-learn 工具包模型使用建议](http://scikit-learn.org/stable/tutorial/machine_learning_map/index.html)：

![]()

**Comment by bighuang624** [comment]
## 特征降维

**特征降维**是无监督学习的一个应用。目的是**重构有效的低维度特征向量**，并**有效表现数据**（使得不因高维度而无法被肉眼观察）。

降维/压缩问题则是选取数据具有代表性的特征，在保持数据多样式（Variance）的基础上，**规避掉大量的特征冗余和噪声**，不过这个过程会**损失一些有用的模式信息**。

经过大量的实践证明，相较于损失的少部分模型性能，维度压缩能过节省大量用于模型训练的时间，综合效率更高。

### 主成分分析

**主成分分析（PCA, Principal Component Analysis）**是最为经典和实用的特征降维技术，特别在辅助图像识别方面有突出的表现。

PCA 将原本的特征空间做了映射，使得新的映射后特征空间数据彼此正交。这样一来，就通过主成分分析尽可能保留下具备区分性的低维数据特征。

**Comment by bighuang624** [comment]
## 特征提升

包括**特征抽取**和**特征筛选**。

* 特征抽取：逐条将原始数据转化为特征向量的形式（涉及对数据特征的量化表示）；
* 特征筛选：在高维度、已量化的特征向量中选择对指定任务更有效的特征组合，进一步提升模型性能。

**Comment by bighuang624** [comment]
## 超参数搜索

由于超参数的空间是无尽的，通常情况下可以依靠**网格搜索（GridSearch）**对多种超参数组合的空间进行暴力搜索。每一套超参数组合被代入到学习函数中作为新的模型，并且为了比较新模型之间的性能，每个模型都会采用交叉验证的方法在多组相同的训练和开发数据集下进行评估。

由于各个新模型在执行交叉验证的过程中是相互独立的，所以可以充分利用多核处理器甚至是分布式的计算资源来从事**并行搜索（Parallel Grid Search）**，这样能够成倍节省运算时间。

**Comment by bighuang624** [comment]
## XGBoost 模型

**提升（Boosting）**分类器隶属于集成学习模型。它的基本思想是把成百上千个分类准确率较低的树模型组合起来，成为一个准确率很高的模型。这个模型的特点在于不断迭代，每次迭代就生成一棵新的树。

对于如何在每一步生成合理的树，有很多方法被提出，例如**梯度提升树（Gradient Tree Boosting）**，在生成每一棵树的时候采用**梯度下降的思想**，以之前生成的所有决策树为基础，向着最小化给定目标函数的方向再进一步。

在合理的参数设置下，往往要生成一定数量的树才能达到令人满意的准确率。在数据集较大较复杂的时候，模型可能需要几千次迭代运算。**XGBoost（全称为 eXtreme Gradient Boosting）**工具可以有效解决这个问题，其最大的特点在于能够自动利用 CPU 的多线程进行并行，且在算法上加以改进提高了精度。
